# Import Packages

In [1]:
import pandas as pd
import numpy as np
import random
from plotnine import *

# repeated printouts
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# datetime 
from datetime import timedelta, date, datetime
from dateutil.relativedelta import relativedelta

# Load the raw data

We are only loading the excel sheets we will use for our research purposes. The sheets we need are `Demographics`, `Presenting Problems`, `Emergency Crisis Services`.

In [2]:
# Read selected excel sheets into pandas dataframes and save in a dictionary
file_path = '../input/Dartmouth Data Set.xlsx'
sheet_names = ['Demographics', 'Presenting Problems', 'Emergency Crisis Services', 'living Situation change', 'In-patient admissions']

# create an empty dictionary to store the sheet data
sheet_data = {}

# loop through each sheet name and read the corresponding sheet data into a pandas dataframe
for sheet_name in sheet_names:
    sheet_data[sheet_name] = pd.read_excel(file_path, sheet_name)

# Save the data in variables and inspect them

In [3]:
# Retrieve the dataframes from the dictionary and save them in variables
demographics = sheet_data['Demographics']
presenting_problems = sheet_data['Presenting Problems']
emergency_crisis_services = sheet_data['Emergency Crisis Services']
living_situation_change = sheet_data['living Situation change']
in_patient_admissions = sheet_data['In-patient admissions']

In [4]:
# Explore the demographics dataframe
demographics.head()
demographics.columns
demographics.info()

Local ID                             Region        County  \
0  8085297     California : CA START East Bay  Contra Costa   
1   440987                New York : Region 3      Saratoga   
2  8008815     California : CA START East Bay       Alameda   
3   220396                New York : Region 3        Albany   
4  6570649  California : CA START San Andreas    Santa Cruz   

     Start Coordinator Date Enrolled in START  \
0        Mary Akhuetie             2020-12-30   
1      Natalia Wolosen             2020-12-29   
2  Sharon Cyrus-Savary             2020-12-29   
3      Matthew Wolcott             2020-12-29   
4        Ilian Navarro             2020-12-29   

                                     Status Status Date  \
0                                    Active  2020-12-30   
1             Inactive (stable functioning)  2021-08-09   
2  Inactive (no longer requesting services)  2021-06-27   
3     Inactive (inappropriate for services)  2021-06-11   
4                                    Active  2020-12-29   

                       Time Enrolled in START  \
0  Business Hours (Monday - Friday 8am - 5pm)   
1  Business Hours (Monday - Friday 8am - 5pm)   
2  Business Hours (Monday - Friday 8am - 5pm)   
3  Business Hours (Monday - Friday 8am - 5pm)   
4  Business Hours (Monday - Friday 8am - 5pm)   

        Source of referral to START Suitability of enrollment in START  ...  \
0  Case Manager/Service Coordinator                        Appropriate  ...   
1  Case Manager/Service Coordinator                        Appropriate  ...   
2  Case Manager/Service Coordinator                        Appropriate  ...   
3  Case Manager/Service Coordinator                        Appropriate  ...   
4  Case Manager/Service Coordinator                        Appropriate  ...   

  State sponsored ICF/DD stay in past 1 - 5 years (at enrollment to START)  \
0                                                 No                         
1                                                 No                         
2                                                 No                         
3                                                 No                         
4                                                 No                         

  Jailed in the past year (at enrollment to START)  \
0                                               No   
1                                               No   
2                                               No   
3                                               No   
4                                               No   

  Jailed in the past 1 - 5 years (at enrollment to START)  \
0                                                 No        
1                                                 No        
2                                                 No        
3                                                 No        
4                                                 No        

  Employment Status at Enrollment Grade In School      Current IEP  \
0                         Student         Grade 7  Yes: 2020-11-16   
1                         Student        Grade 11               No   
2                         Student         Grade 8  Yes: 2020-01-01   
3                    Not employed             NaN              NaN   
4                             NaN             NaN              NaN   

  Current 504 Plan                                School Setting  \
0               No                Special school-out of district   
1               No                            Other: BOCES 4:1:2   
2               No                    Special school-in district   
3              NaN                                           NaN   
4              NaN  Other: COE Program Held at Cabrillo College    

     Custody Status Child Lives With  
0  Parental custody          Parents  
1  Parental custody          Parents  
2  Parental custody          Parents  
3               NaN              NaN  
4               N

Index(['Local ID', 'Region', 'County', 'Start Coordinator',
       'Date Enrolled in START', 'Status', 'Status Date',
       'Time Enrolled in START', 'Source of referral to START',
       'Suitability of enrollment in START',
       'Individual/Caregiver reliable access to technology',
       'Presenting problems at time of enrollment', 'Services at Enrollment',
       'Date of birth', 'Gender', 'Race', 'Ethnicity',
       'Level of Intellectual Disability',
       'Current living situation at enrollment to START',
       'More than one living situation over the last 5 years?',
       'More than one living situation over the last year?',
       'Psychiatric hospitalization in past year (at enrollment to START)',
       'Psychiatric hospitalizations in past 1 - 5 years (at enrollment to START)',
       'Emergency department visits in past year (at enrollment to START)',
       'Emergency department visits in past 1-5 years (at enrollment to START)',
       'Law enforcement encounters i

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4986 entries, 0 to 4985
Data columns (total 48 columns):
 #   Column                                                                                  Non-Null Count  Dtype         
---  ------                                                                                  --------------  -----         
 0   Local ID                                                                                4986 non-null   object        
 1   Region                                                                                  4986 non-null   object        
 2   County                                                                                  4986 non-null   object        
 3   Start Coordinator                                                                       4947 non-null   object        
 4   Date Enrolled in START                                                                  4986 non-null   datetime64[ns]
 5   Status              

In [5]:
# Explore the emergency crisis services dataframe
emergency_crisis_services.head()
emergency_crisis_services.columns
emergency_crisis_services.info()

Individual ID                                         Region  \
0      11153125  New Hampshire : Region 2 - Pathways/Claremont   
1      11153125  New Hampshire : Region 2 - Pathways/Claremont   
2      11153125  New Hampshire : Region 2 - Pathways/Claremont   
3      11158347  New Hampshire : Region 9 - Community Partners   
4      11158347  New Hampshire : Region 9 - Community Partners   

  Date/Time of Contact                 Source of Contact  \
0  2016-08-16 08:40:00  Case Manager/Service Coordinator   
1  2018-04-27 15:00:00  Case Manager/Service Coordinator   
2  2018-06-05 16:00:00  Residential provider - Community   
3  2015-06-01 14:00:00                     Family member   
4  2015-06-01 16:00:00                     Family member   

                                  Reason for Contact  \
0  At risk of losing placement, Other, Self-injur...   
1             Mental health symptoms, Self-injurious   
2  Mental health symptoms, Suicidal ideation/beha...   
3  Diagnosis and treatment plan assistance, Other...   
4  Aggression (physical, verbal, property destruc...   

  Type of Emergency Response  \
0    Phone consultation only   
1  In-person: Emergency room   
2     Other: phone and email   
3    Phone consultation only   
4    Phone consultation only   

   Total episode length (initial call to disposition)  \
0                                               15.0    
1                                               60.0    
2                                              120.0    
3                                              254.0    
4                                               70.0    

   Total response time (initial call to arrival on-site)  Travel Time  \
0                                                NaN              NaN   
1                                              240.0             60.0   
2                                                NaN              NaN   
3                                                NaN              NaN   
4                                                NaN              NaN   

  Police Response  ... Restraints Used If Restraints Were Used  \
0              No  ...              No                     NaN   
1              No  ...              No                     NaN   
2              No  ...              No                     NaN   
3              No  ...              No                     NaN   
4              No  ...              No                     NaN   

                                 Disposition  \
0                  Other: outpatient MH appt   
1  START emergency resource center admission   
2             Psychiatric Hospital Admission   
3  START emergency resource center admission   
4                   Maintain current setting   

            Disposition Information Admitted through ED/CPEP  \
0                               NaN                       No   
1                               NaN                       No   
2                               NaN                       No   
3                               NaN                       No   
4  Planned Resource Center referral                       No   

  Date of Admission Disposition Information (ER)  \
0               NaT                          NaN   
1               NaT                          NaN   
2               NaT                          NaN   
3               NaT                          NaN   
4               NaT                          NaN   

  Did START response prevent higher level of care (hospital diversion)  \
0                                                 No                     
1                                                Yes                     
2                                                Yes                     
3                                                 No                     
4                                                 No                     

  RSQ Completed?         Entered By  
0             No  Valarie Tetreault  
1             No   Samantha

Index(['Individual ID', 'Region', 'Date/Time of Contact', 'Source of Contact',
       'Reason for Contact', 'Type of Emergency Response',
       'Total episode length (initial call to disposition)',
       'Total response time (initial call to arrival on-site)', 'Travel Time',
       'Police Response', 'If Police Responded', 'Who Contacted Police?',
       'Mobile Crisis', 'If Mobile Crisis', 'Restraints Used',
       'If Restraints Were Used', 'Disposition', 'Disposition Information',
       'Admitted through ED/CPEP', 'Date of Admission',
       'Disposition Information (ER)',
       'Did START response prevent higher level of care (hospital diversion)',
       'RSQ Completed?', 'Entered By'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18168 entries, 0 to 18167
Data columns (total 24 columns):
 #   Column                                                                Non-Null Count  Dtype         
---  ------                                                                --------------  -----         
 0   Individual ID                                                         18168 non-null  object        
 1   Region                                                                18168 non-null  object        
 2   Date/Time of Contact                                                  18168 non-null  datetime64[ns]
 3   Source of Contact                                                     17780 non-null  object        
 4   Reason for Contact                                                    17983 non-null  object        
 5   Type of Emergency Response                                            17930 non-null  object        
 6   Total episode length (initial call to 

In [6]:
# Explore the presenting problems dataframe
presenting_problems.head()
presenting_problems.columns
presenting_problems.info()

Client Local Id START Coordinator                         Region  \
0               1     Ashley Gruver  Iowa : County Social Services   
1               1     Ashley Gruver  Iowa : County Social Services   
2               1     Ashley Gruver  Iowa : County Social Services   
3         000083W          Amy Huff          North Carolina : West   
4         000083W          Amy Huff          North Carolina : West   

                   Presenting Problems at Enrollment  
0  Aggression (physical, verbal, property destruc...  
1                             Mental health symptoms  
2                                    Self-injurious   
3  Aggression (physical, verbal, property destruc...  
4  Decrease in ability to participate in daily fu...

Index(['Client Local Id', 'START Coordinator', 'Region',
       'Presenting Problems at Enrollment'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31014 entries, 0 to 31013
Data columns (total 4 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   Client Local Id                    31014 non-null  object
 1   START Coordinator                  29435 non-null  object
 2   Region                             31014 non-null  object
 3   Presenting Problems at Enrollment  31014 non-null  object
dtypes: object(4)
memory usage: 969.3+ KB


Let's skim `Presenting Problems at Enrollment` to see the presented problems

In [7]:
random.choices(presenting_problems['Presenting Problems at Enrollment'].unique(), k = 20)

['Other: Lighting things on fire',
 'Other: Property Destruction',
 'Other: Redirection Guidance',
 'Other: Found his parents deceased in the home 10/2015',
 'Other: changes in weather (does not like AC/fan in his personal space)',
 'Other: IRA/Day Hab Staff needs assistance',
 'Other: recent hospitalization 2/2021',
 'Other: Does not want to leave the home',
 'Other: Drug Use; Stealing',
 'Other: Training needed for providers.',
 'Other: History of anal stimulation',
 'Other: Urinating on classroom floor',
 'Other: Residential provider needs assistance with the transition from residential school',
 'Other: Does not sleep at night',
 'Other: refusal to participate in services',
 'Other: refusal behavior',
 'Other: Recent hospitalization; Cannot achieve stability',
 'Other: PICA',
 'Other: runny away from group home',
 'Other: Complex medical needs']

<font color = "green"> As we can see above the presented problems are all over the place. In the next section, we will focus on cleaning this column using regex to narrow down the presented problems.  </font>

# Data cleaning/preprocessing

Let's first `Presenting Problems at Enrollment` column to `presented_problems` for more readability.

In [8]:
# rename the column
presenting_problems.rename(columns={'Presenting Problems at Enrollment': 'presented_problems'}, inplace=True)

# confirm that the column was renamed successfully
presenting_problems.columns

Index(['Client Local Id', 'START Coordinator', 'Region', 'presented_problems'], dtype='object')

In [9]:
presenting_problems['presented_problems'].value_counts()

Aggression (physical, verbal, property destruction, threats)                                                                                        7427
Family needs assistance                                                                                                                             4489
Mental health symptoms                                                                                                                              4356
Self-injurious                                                                                                                                      2640
Diagnosis and treatment plan assistance                                                                                                             2246
                                                                                                                                                    ... 
Other: Social Skills                                                              

Now, let's turn all the presenting problems in lowercase to see add consistency.

In [10]:
presenting_problems['presented_problems'] = presenting_problems['presented_problems'].str.lower()
presenting_problems.sample(2)

Client Local Id START Coordinator  \
28791         8143879     Sydney Myhers   
20160          413950   Marie Toussaint   

                                              Region    presented_problems  
28791              California : CA START San Andreas  leaving unexpectedly  
20160  New York : Region 4 Richmond Kings - Brooklyn  leaving unexpectedly

In [11]:
# subset to rows that do not contain 'other:' for presented problems
presenting_problems_no_other = presenting_problems[~presenting_problems["presented_problems"].str.contains('other:')].copy()
presenting_problems_no_other.presented_problems.unique()

presenting_problems_no_other.shape[0]

print(f"The number of unique presenting problems before dropping 'Other' is: {presenting_problems.presented_problems.nunique()}")
print(f"The number of unique presenting problems is: {presenting_problems_no_other.presented_problems.nunique()}")

array(['aggression (physical, verbal, property destruction, threats)',
       'mental health symptoms', 'self-injurious ',
       'decrease in ability to participate in daily functions',
       'leaving unexpectedly', 'diagnosis and treatment plan assistance',
       'family needs assistance', 'at risk of losing placement',
       'sexualized behavior', 'suicidal ideation',
       'transition from hospital', 'suicidal action'], dtype=object)

30011

The number of unique presenting problems before dropping 'Other' is: 704
The number of unique presenting problems is: 12


Let's see how many rows we are losing by removing presented problems that start with `Other:`

In [12]:
print(f"Retention Rate After Dropping 'Other': {round(100*presenting_problems_no_other.shape[0]/presenting_problems.shape[0], 2)}%")

Retention Rate After Dropping 'Other': 96.77%


A retention rate of 96.77% is really good, so we will stick with the 12 unique presented problems. Now, we will write a function to parse the `Region` variable to `state` and `region`

In [13]:
def parse_region(row):
    """
    Parse a string in the format "<state>: <region>" and return a tuple
    containing the state and region.
    
    Parameters:
    row (str): A string in the format "<state> : <region>".
    
    Returns:
    A tuple containing the parsed state and region.
    """
    state, region = row.lower().split(':')
    return state.strip(), region.strip()

# apply the function on the demographics, presenting_problems_no_other, and emergency_crisis_services dataframes
presenting_problems_no_other.loc[:, ['state', 'region_new']] = presenting_problems_no_other['Region'].apply(parse_region).tolist()
demographics.loc[:, ['state', 'region_new']] = demographics['Region'].apply(parse_region).tolist()
emergency_crisis_services.loc[:, ['state', 'region_new']] = emergency_crisis_services['Region'].apply(parse_region).tolist()

# sample from the presenting_problems_no_other dataframe
presenting_problems_no_other.sample(3)

Client Local Id  START Coordinator                             Region  \
26562         6597201    Megan Hernandez  California : CA START San Andreas   
10062          236323    Andrea Petersen                New York : Region 3   
13592          316874  Ramona Policastro                New York : Region 3   

                                      presented_problems       state  \
26562  decrease in ability to participate in daily fu...  california   
10062  decrease in ability to participate in daily fu...    new york   
13592  decrease in ability to participate in daily fu...    new york   

                 region_new  
26562  ca start san andreas  
10062              region 3  
13592              region 3

In [14]:
# check all the unique states in the presenting_problems_no_other dataframe
presenting_problems_no_other.state.unique()

array(['iowa', 'north carolina', 'texas', 'new york', 'california',
       'colorado', 'new hampshire', 'maryland'], dtype=object)

Next, we will merge the `demographic` and `presenting_problems_no_other` dataframes using the `Client ID`.

In [15]:
demographics_problems_merged = presenting_problems_no_other.merge(demographics, left_on='Client Local Id', right_on='Local ID',
                                                                  how='inner', suffixes=('_presenting_problems', '_demographics'), 
                                                                  indicator = True).drop_duplicates()

In [16]:
demographics_problems_merged.sample(3)
demographics_problems_merged.shape

Client Local Id  START Coordinator Region_presenting_problems  \
12707          436376        Sarah Guess        New York : Region 1   
9519           372564  Rebecca Szymanski        New York : Region 1   
81            1003676      Jamie Edwards     Texas : Tarrant County   

                                      presented_problems  \
12707                            family needs assistance   
9519                             family needs assistance   
81     aggression (physical, verbal, property destruc...   

      state_presenting_problems region_new_presenting_problems Local ID  \
12707                  new york                       region 1   436376   
9519                   new york                       region 1   372564   
81                        texas                 tarrant county  1003676   

          Region_demographics   County  Start Coordinator  ...  \
12707     New York : Region 1   Monroe        Sarah Guess  ...   
9519      New York : Region 1     Erie  Rebecca Szymanski  ...   
81     Texas : Tarrant County  Tarrant      Jamie Edwards  ...   

      Employment Status at Enrollment Grade In School      Current IEP  \
12707                    Not employed         Grade 7  Yes: 2018-09-09   
9519                     Not employed         Grade 8  Yes: 1969-12-31   
81                       Not employed             NaN              NaN   

      Current 504 Plan                         School Setting  \
12707               No                       Home instruction   
9519                No  Other: Separate special ed classroom    
81                 NaN                                    NaN   

         Custody Status   Child Lives With state_demographics  \
12707  Relative custody          Relatives           new york   
9519   Parental custody            Parents           new york   
81                  NaN  Residential staff              texas   

      region_new_demographics _merge  
12707                region 1   both  
9519                 region 1   both  
81             tarrant county   both  

[3 rows x 57 columns]

(17808, 57)

Using the `Date Enrolled in START` variable in the merged data frame, we will make a binary classifier of whether or not the client enrolled pre or post COVID. We will use March 01, 2020 as the cutoff date.

In [17]:
# convert to datetime format and save in enrollment_date
demographics_problems_merged['enrollment_date'] = pd.to_datetime(demographics_problems_merged['Date Enrolled in START'])
print(f"Data type for the date variable: {demographics_problems_merged['enrollment_date'].dtype}")

# create a binary clarity column based on pre and post covid.
date_cutoff = pd.to_datetime("2020-03-01")
demographics_problems_merged['is_post_covid'] = demographics_problems_merged['enrollment_date'] >= date_cutoff

# check the value counts to see how many of the inputs are post covid
demographics_problems_merged['is_post_covid'].value_counts()

Data type for the date variable: datetime64[ns]


False    14270
True      3538
Name: is_post_covid, dtype: int64